File: Cloud Storage System/Level 3/C#/file_repository.cs

// FileRepository.cs
using System.Collections.Generic;

public class FileRepository
{
    private readonly Dictionary<string, (File file, string owner)> _storage = new();

    public bool Exists(string path)
    {
        return _storage.ContainsKey(path);
    }

    public bool Add(File file, string owner)
    {
        if (Exists(file.Path)) return false;
        _storage[file.Path] = (file, owner);
        return true;
    }

    public File? Get(string path)
    {
        return _storage.TryGetValue(path, out var entry) ? entry.file : null;
    }

    public string? GetOwner(string path)
    {
        return _storage.TryGetValue(path, out var entry) ? entry.owner : null;
    }

    public (File, string)? Delete(string path)
    {
        if (!_storage.TryGetValue(path, out var entry)) return null;
        _storage.Remove(path);
        return entry;
    }

    public List<(File, string)> GetAll()
    {
        return new List<(File, string)>(_storage.Values);
    }
}

File: Cloud Storage System/Level 3/Python/file_repository.py
# file_repository.py

from file import File

class FileRepository:
    def __init__(self):
        self._storage = {}  # path -> (File, owner_user_id)

    def exists(self, path: str) -> bool:
        return path in self._storage

    def add(self, file: File, user_id: str) -> bool:
        if self.exists(file.path):
            return False
        self._storage[file.path] = (file, user_id)
        return True

    def get(self, path: str) -> File | None:
        entry = self._storage.get(path)
        return entry[0] if entry else None

    def get_owner(self, path: str) -> str | None:
        entry = self._storage.get(path)
        return entry[1] if entry else None

    def delete(self, path: str) -> tuple[File, str] | None:
        return self._storage.pop(path, None)

    def get_all(self) -> list[tuple[File, str]]:
        return list(self._storage.values())
